pkg/scheduler/framework/genericpreemption/preemption.go

```diff
@@ -0,0 +1,514 @@
+	// GetOffsetAndNumCandidates chooses a random offset and calculates the number of candidates that should be
+	// shortlisted for dry running preemption.
+	GetOffsetAndNumCandidates(nodes int32) (int32, int32)
+	// CandidatesToVictimsMap builds a map from the target node to a list of to-be-preempted Pods and the number of PDB violation.
+	CandidatesToVictimsMap(candidates []Candidate) map[string]*extenderv1.Victims
+	// DryRunPreemption simulates Preemption logic on target node, returns preemption candidates and a map indicating filtered nodes statuses.
+	DryRunPreemption(context.Context, framework.Handle, *framework.CycleState, *v1.Pod, []*framework.NodeInfo, []*policy.PodDisruptionBudget, int32, int32) ([]Candidate, framework.NodeToStatusMap)
+}
+
+type PreemptionEvaluator struct {
+	Name      string
+	Handler   framework.Handle
+	PodLister corelisters.PodLister
+	PdbLister policylisters.PodDisruptionBudgetLister
+	PreemptionInteface
+}
+
+func (ev *PreemptionEvaluator) Preempt(ctx context.Context, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, *framework.Status) {
+	cs := ev.Handler.ClientSet()
+	nodeLister := ev.Handler.SnapshotSharedLister().NodeInfos()
+
+	// 0) Fetch the latest version of <pod>.
+	// It's safe to directly fetch pod here. Because the informer cache has already been
+	// initialized when creating the Scheduler obj, i.e., factory.go#MakeDefaultErrorFunc().
+	// However, tests may need to manually initialize the shared pod informer.
+	podNamespace, podName := pod.Namespace, pod.Name
+	pod, err := ev.PodLister.Pods(pod.Namespace).Get(pod.Name)
+	if err != nil {
+		klog.ErrorS(err, "getting the updated preemptor pod object", "pod", klog.KRef(podNamespace, podName))
+		return "", framework.AsStatus(err)
+	}
+
+	// 1) Ensure the preemptor is eligible to preempt other pods.
+	if !PodEligibleToPreemptOthers(pod, nodeLister, m[pod.Status.NominatedNodeName]) {
+		klog.V(5).InfoS("Pod is not eligible for more preemption", "pod", klog.KObj(pod))
+		return "", nil
+	}
+
+	// 2) Find all preemption candidates.
+	candidates, nodeToStatusMap, status := ev.FindCandidates(ctx, state, pod, m)
+	if !status.IsSuccess() {
+		return "", status
+	}
+
+	// Return a FitError only when there are no candidates that fit the pod.
+	if len(candidates) == 0 {
+		fitError := &framework.FitError{
+			Pod:         pod,
+			NumAllNodes: len(nodeToStatusMap),
+			Diagnosis: framework.Diagnosis{
+				NodeToStatusMap: nodeToStatusMap,
+				// Leave FailedPlugins as nil as it won't be used on moving Pods.
+			},
+		}
+		return "", framework.NewStatus(framework.Unschedulable, fitError.Error())
+	}
+
+	// 3) Interact with registered Extenders to filter out some candidates if needed.
+	candidates, status = ev.CallExtenders(ev.Handler.Extenders(), pod, nodeLister, candidates)
+	if !status.IsSuccess() {
+		return "", status
+	}
+
+	// 4) Find the best candidate.
+	bestCandidate := ev.SelectCandidate(candidates)
```

Comment:
Hmm... SelectCandidate() is not in the interface, so it implies it's not able to be customized. So are you going to make it customizable or not? If yes, make it a function of the `Interface`, otherwise make it private.

Reply:
This is debatable though,  I have made `CandidatesToVictimsMap` as a interface func (debatable too), and it's likely `pickOneNodeForPreemption` could be overwritten as well? maybe we can keep it as is and make it a interface when there is a real case that need to overwrite this func?